File: 백준/Gold/9663. N－Queen/N－Queen.py
import sys


def is_promising(x):
    global g
    for i in range(x):
        if g[i] == g[x] or abs(g[i] - g[x]) == abs(x - i):
            return False
    return True


def n_queen(x):
    global ans
    if x == n:
        ans += 1
        return
    for i in range(n):
        global g
        g[x] = i
        if is_promising(x):
            n_queen(x + 1)


n = int(sys.stdin.readline().rstrip())
ans = 0
g = [0 for _ in range(n)]
n_queen(0)
print(ans)

File: 백준/Gold/9663. N－Queen/N－Queen.cc
#include <bits/stdc++.h>
using namespace std;
int n, g[16], ans;
bool isPromising(int x) {
    for (int i = 0; i < x; i++) {
        if (g[x] == g[i] || abs(g[x] - g[i]) == abs(x - i)){
            return false;
        }
    }
    return true;
}

void nQueen(int x) {
    if (x == n) {
        ans += 1;
        return;
    }
    for (int i = 0; i < n; i++) {
        g[x] = i;
        if (isPromising(x)) {
            nQueen(x + 1);
        }
    }
}

int main() {
    cin >> n;
    nQueen(0);
    cout << ans;
    return 0;
}